Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs

```diff
@@ -51,15 +93,29 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
         if (GetBallisticShots(component) >= component.Capacity)
             return;
 
-        component.Entities.Add(args.Used);
-        Containers.Insert(args.Used, component.Container);
-        // Not predicted so
-        Audio.PlayPredicted(component.SoundInsert, uid, args.User);
+        TimeSpan insertDelayConverted;
+
+        if (component.InsertDelay > 0)
+        {
+            insertDelayConverted = TimeSpan.FromSeconds(component.InsertDelay);
+            _doAfter.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, insertDelayConverted, new DelayedAmmoInsertDoAfterEvent(), used: args.Used, target: args.Target, eventTarget: uid)
+            {
+                BreakOnMove = true,
+                BreakOnDamage = false,
+                NeedHand = true
+            });
+        }
+        else // If there's no custom InsertDelay on this component, immediately load the ammo in. Shotguns and mag-filling.
+        {
+            ManualLoad(uid, component, args.Used, args.User);
+        }
         args.Handled = true;
-        UpdateBallisticAppearance(uid, component);
-        Dirty(uid, component);
     }
 
+    /// <summary>
+    /// Interacting with a BallisticAmmoProvider with another one, to transfer ammo.
+    /// Uses FillDelay, defaulting to 0.5s
+    /// </summary>
```

Comment:
If it's upstream code not commenting it or commenting it with an upstream PR is better, for conflicts